libs/saml2-web-sso/src/SAML2/WebSSO/API.hs

```diff
@@ -95,42 +98,66 @@ defResponseURI = getSsoURI (Proxy @API) (Proxy @APIAuthResp')
     authnResponseBodyRaw :: MultipartData Mem
+    -- FUTUREWORK: this is only for dumping the error on the "something went wrong" page.  we
+    -- should find a better solution there and remove it here.
   }
 
 renderAuthnResponseBody :: AuthnResponse -> LBS
 renderAuthnResponseBody = EL.encode . cs . encode
 
--- | Implies verification, hence the constraints.
+-- | Implies verification, hence the constraints and the optional service provider ID (needed for IdP lookup).
 parseAuthnResponseBody ::
   forall m err spid extra.
-  (SPStoreIdP (Error err) m, spid ~ IdPConfigSPId m, extra ~ IdPConfigExtra m) =>
+  (SPStoreIdP (Error err) m, SPStoreRequest AuthnRequest m, spid ~ IdPConfigSPId m, extra ~ IdPConfigExtra m) =>
   Maybe spid ->
   ST ->
-  m (AuthnResponse, IdPConfig extra)
+  m (NonEmpty Assertion, IdPConfig extra, UnvalidatedSAMLStatus)
 parseAuthnResponseBody mbSPId base64 = do
   -- https://www.ietf.org/rfc/rfc4648.txt states that all "noise" characters should be rejected
   -- unless another standard says they should be ignored.  'EL.decodeLenient' chooses the radical
   -- approach and ignores all "noise" characters.  since we have to deal with at least %0a, %0d%0a,
   -- '=', and probably other noise, this seems the safe thing to do.  It is no less secure than
   -- rejecting some noise characters and ignoring others.
   let xmltxt :: LBS = EL.decodeLenient (cs base64 :: LBS)
-  resp <-
-    either (throwError . BadSamlResponseXmlError . cs) pure $
-      decode (cs xmltxt)
-  issuer <- maybe (throwError BadSamlResponseIssuerMissing) pure (resp ^. rspIssuer)
-  idp :: IdPConfig extra <- getIdPConfigByIssuerOptionalSPId issuer mbSPId
-  creds <- idpToCreds issuer idp
-  simpleVerifyAuthnResponse creds xmltxt
-  pure (resp, idp)
+
+  (signedAssertions, idp, status) <- do
+    resp <-
+      -- do not use `resp` as a result of `parseAuthnResponseBody`!  only use what comes back
+      -- from `simpleVerifyAuthnResponse`!
+      either (throwError . BadSamlResponseXmlError . cs) pure $
+        decode @_ @AuthnResponse (cs xmltxt)
+    issuer <- do
+      respIssuer :: Issuer <-
+        -- this issuer is not signed!!  but we'll check it anyway, just for good measure and
+        -- because the standard says so.
+        maybe (throwError BadSamlResponseIssuerMissing) pure (resp ^. rspIssuer)
+      signedIssuers :: NonEmpty Issuer <-
+        -- these are *possibly* signed, but we collect all of them, and if none of them are
+        -- signed, signature validation will fail later.
+        pure (view assIssuer <$> resp ^. rspPayload)
+      case L.nub (respIssuer : toList signedIssuers) of
+        [i] -> pure i
+        _ -> throwError BadSamlResponseInconsistentIdPIssuerInfo
+
+    idp :: IdPConfig extra <-
+      -- this is convoluted, but secure against signatures from rogue idps: this idp config is
+      -- (a) is signed by the given in the authentiation response issuer, and
```

Comment:
```suggestion
      -- (a) is signed by the given certificate in the authentiation response issuer, and
```
Missing word? :thinking:

Reply:
i rewrote it a little more.  how about now?